### DISCUSSION TITLE
MKS Serial protocol

### DISCUSSION THREAD

**Discussion by luc-github** [luc-github/ESP3D#620] [F.A.Q]
Some notes on MKS serial :
0 - be sure  it is enabled `#define COMMUNICATION_PROTOCOL MKS_SERIAL `
1 - ESP3D baud rate must be 115200 as it is MKS specs
2 - Several commands are filtered by TFT / Printer MKS code 
so Marlin printer panel may not work - at least today, this is out of scope of ESP3D but feel free to list here what commands are not working, so it could help MKS team or anyone willing to work on it to check the fix/workaround is working
3 - You do not need to configure in Marlin the 2nd port on MKS Robin, this is handled by Marlin MKS code, doing it may bring some trouble in communication
4 - if you are not using MKS ESP boards but nodeMCU or whatever different board, you may experienced some upload failures, this is due to that upload speed is taillored to MKS devices, other board may be more sensitive to serial noise (EMI), so you may need to decrease the upload speed in Marlin Mks code and ESP3D code to compensate, which of course will affect upload speed...

**Comment by ColdBeer72** [comment]
I had everything configured just like your instructions for my Robin Nano V3 and MKS Wifi board. Everything at 115200bps (both Marlin and the compiled ESP3D). I tried with 2nd port but disabled it with this FAQ. But still no communication between the wifi board and the mainboard. Webdav disabled... But still can't communicate both boards. Any possible test?

**Reply by ColdBeer72** [reply]
To be clear, I had some kind of communication some days ago (I could see sd card content, or start fan...) but never could check marlin parameters. Anyway, at the moment I'm totally stopped because there is no communication at all.

**Reply by Freddyb888** [reply]
I have the same problem now. Do t know what to do, flashed it 3 times already

**Comment by luc-github** [comment]
Marlin parameters is not working is kind of `normal` it is MKS limitation as mentioned above

Nothing changed between when working and not working ? FW update ? hardware change ? power cut ?

**Reply by ColdBeer72** [reply]
Lots of firmware updates :-D :-D :-D But nothing really important. This is a new board and I'm in the process of fine tunning (TMC drivers, speed, acceleration, bltouch...).
So, basically, with MKS and ESP3D, there is not so much to do with the board?

**Comment by luc-github** [comment]
if many fw update and before was working -> it is user issue using wrong setting somewhere

**Comment by luc-github** [comment]
> MKS and ESP3D, there is not so much to do with the board?

What do you mean ?

**Reply by ColdBeer72** [reply]
Most interesting use of ESP3D is to notify (via Telegram, for example) of the end of a job... or change configuration options via gcode (like pronterface does)... if I can't work with Marlin parameteres nor check the machine status... What could I still do?

**Reply by luc-github** [reply]
you can send all commands - the problem is the Mks code filter the answer currently - it is a bug - I have reported them the issue - I have no time to  fix by  myself and it is in Marlin code
you can control printer all commands are sent
upload files is working with speed 110KB/S
send print command is working
monitoring temperature is working

to sum up everything the MKS plugin do is working

the problem is the monitoring of some controls /printer feedback like M503, fan, speed, etc...  but I did not check which commands are not answering as it is out of my scope

If you are not interrested in fast file upload - then do not use MKS protocol and use standard  serial and everything will be working but fast upload, I plan to work on better protocol (https://github.com/luc-github/ESP3D/issues/575) but I am currently on webUI for 3.0

**Reply by ColdBeer72** [reply]
Sorry, after a few days of testing...

1. Temperature monitoring is not working
2. Upload files works
3. Send print command works (but no answer in any way from Marlin)
4. Every other thing which means to receive data from Marlin does not work.

So M118 to send a notification doesn't work too.
You talk about a code filter in MKS but... Marlin MKS is pretty near nowadays to standard Marlin. So... where is the filter? Is a motherboard filter? A filter inside MKS-WIFI module? A filter into mks's marlin? Me, with others, wanna help and trying to help to improve this, but it should go faster if you give us some indications about your knowledge of the problem.

Thanks.

**Reply by luc-github** [reply]
the MKS code in marlin is here:
https://github.com/MarlinFirmware/Marlin/tree/2.0.x/Marlin/src/lcd/extui/lib/mks_ui
the filter is here : https://github.com/MarlinFirmware/Marlin/blob/2.0.x/Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp#L595
it does not do a real bridge
I did not digged more how to solve it as it is out of my scope and I am already busy with my projects

About temperatures it is weird because this one is sent by MKS code, so you may have another problem if M105  or even M115 commands do not send any data to ESP3D

**Reply by themacboy** [reply]
> you can send all commands - the problem is the Mks code filter the answer currently - it is a bug - I have reported them the issue - I have no time to fix by myself and it is in Marlin code
> you can control printer all commands are sent
> upload files is working with speed 110KB/S
> send print command is working
> monitoring temperature is working
> 
> to sum up everything the MKS plugin do is working
> 
> the problem is the monitoring of some controls /printer feedback like M503, fan, speed, etc... but I did not check which commands are not answering as it is out of my scope
> 
> If you are not interrested in fast file upload - then do not use MKS protocol and use standard serial and everything will be working but fast upload, I plan to work on better protocol (#575) but I am currently on webUI for 3.0

Ohhh Let me know where is the issue reported and I will add more message requesting a patch ;) hehehe

**Reply by luc-github** [reply]
no issue number - I was discussing directly with MKS team members - I am in touch with them

**Reply by ColdBeer72** [reply]
So, then the "filter" just happens when you use LVGL interface. Then, with Marlin UI there should be no problem. Isn't it?

**Reply by luc-github** [reply]
without this code no mks protocole, so no fast upload it is standard serial coms

**Reply by ColdBeer72** [reply]
So, with LVGL UI fast transfer speed but no control, and with native Marlin UI slow transfers but full control. Is that right?

**Reply by luc-github** [reply]
well I have control with mks code but not all feedback : sd list is working temperatures are working  jog is working

**Reply by ColdBeer72** [reply]
Sorry Luc, but I think that's not correct: **temps are not working**... and I'm not sure what do you want to say telling "jog". For me, just SD and order printing is working. I could send gcodes through terminal but not having feedback (the answer for any codes) so atm we can't say we're having "partial feedback", just no feedback at all.

**Reply by luc-github** [reply]
I mean on my system :
M20  / M105 / M115 give feedback 
Jog is jogging commands to move extruder
so if  on your side  you have zero feedback from Marlin you may have another issue (Marlin configuration? connection issue between MKS TX and ESP RX  ?)

What output in WebUI terminal for :
M105
M20
M115

Edit: I used https://github.com/makerbase-mks/Mks-Robin-Nano-Marlin2.0-Firmware not https://github.com/MarlinFirmware/Marlin, additionally it could be also a broken code in Marlin since I ported the code

**Reply by ColdBeer72** [reply]
This is my output for those commands (verbose mode on):
```
> M20
Begin file list
TSD9C3~5.GCO
TSD99D~5.GCO
RET.DIR
TST40-~9.GCO
TST40-~8.GCO
TST40-~7.GCO
TST40-~6.GCO
TST40-~5.GCO
TESTS.DIR
TST40-~4.GCO
TST40-~3.GCO
TST40-~2.GCO
TST40-~1.GCO
End file list
ok
> M105
ok
T: / B: / T0: / T1: / @:0 B@:0
> M115
ok
FIRMWARE_NAME:Robin_nano
```
The answer for M105 is false, totally false, because printer is at this moment preheated (and finished) for PLA.

If "jog" is moving headset/nozzle/hotend/extruder, then YES, it moves, but sincerely, that's the less interesting thing at all.

I supposed you were talking about the MKS's Marlin version but as LVGL code in official Marlin is developed by MKS, the "filter" is on both versions. Anyway, we have a bit step further and could say that feedback from gcodes are present, so a little step (actually, btw, there is no "upload" button on files section).

**Reply by luc-github** [reply]
```
> M20
Begin file list
TSD9C3~5.GCO
TSD99D~5.GCO
RET.DIR
TST40-~9.GCO
TST40-~8.GCO
TST40-~7.GCO
TST40-~6.GCO
TST40-~5.GCO
TESTS.DIR
TST40-~4.GCO
TST40-~3.GCO
TST40-~2.GCO
TST40-~1.GCO
End file list
ok
> M105
ok
T: / B: / T0: / T1: / @:0 B@:0
> M115
ok
FIRMWARE_NAME:Robin_nano
```
So communication is working - Temperature issue is on Marlin side not ESP3D, cannot show what is not presented 

 
 >there is no "upload" button on files section  
 
 You should have:
 
![image](https://user-images.githubusercontent.com/8822552/117494709-fff14080-af74-11eb-8a4c-079025a10653.png)

 
 Did you enabled Printer SD in WebUI  settings ?

**Comment by themacboy** [comment]
I just installed correctly all, and I get good feedback on my Mks Robin
Nano v.3.

I'm not using MKS ui (LVGL), I'm using Marlin Color UI.

Unfortunately I can't upload files on my sd card or USB driver. Upload icon
is missing when selected.

**Reply by luc-github** [reply]
So you do not use MKS protocol but standard serial protocol  : https://github.com/luc-github/ESP3D/discussions/540#discussioncomment-191759
where Upload is not implemented

**Reply by Legsmaniac** [reply]
> Unfortunately I can't upload files on my sd card

I really wouldn't bother with trying to upload files to SD via ESP3D, it's SOOOOOOO slow.
I bought and use a FlashAir WiFi wireless SD card instead.  MUCH faster and more reliable.

**Reply by ColdBeer72** [reply]
> So you do not use MKS protocol but standard serial protocol : [#540 (comment)](https://github.com/luc-github/ESP3D/discussions/540#discussioncomment-191759)
> where Upload is not implemented

Ok, thanks a lot Luc, for your answers and the whole work you're doing. Bit thanks!!!

BTW, I think you could imagine we (MKS board owners) are trying to make an idea of what we could do with ESP3D in both circunstances: MKS-Marlin (LVGL or Marlin UI) and pure-Marlin compilation (both options: LVGL-Marlin UI).

For what I could understand:

1. **Marlin UI** has no upload options, but complete communication (input/output), so we could play with notifications and so on.
2. In the other side, **LVGL UI** has fast upload, but just output (send G-code but not feedback) from ESP3D to the motherboard

Is that correct? Is that the actual situation?

We are really excited with this, but want to explain it right to our group, to despite false expectations. I'm pretty sure you understand.

**Reply by luc-github** [reply]
1 - I am sorry but I do not know Marlin UI, what is that ? another flavor or Marlin ? if Marlin UI has no support of MKS protocol then yes,  no upload for ESP3 3.0, but 2.1 has upload.
I only used the MKS Marlin version as it was a reference for the Robin V2 I got from MKS, so used default settings.

2 - limited feedback on my side, not `no feedback at all` originally protocol was used with https://github.com/Jeredian/mks-wifi-plugin so Temperatures / SD content /  and send GCODE was enough, but ESP3D need more for full features like for notifications, positions, EEPROM, etc...

**Reply by ColdBeer72** [reply]
I mean, for your clarification:
1- you could download and compile Marlin from Marlin's repository or from MKS's repository. I think both share the same libraries for MKS wifi nowadays.
2-in configuration.h you could select which UI do you want to use (MKS -lvgl-, Marlin Color UI or Marlin TFT -just like a classic TFT screen). Depending which UI you select, you could see that LVGL has a different control of things, like MKS wifi.

In my case, with LVGL UI, I have no control at all, no temps, no feedback from commands, just fast transfers.

**Reply by luc-github** [reply]
as mentionned you use Marlin repository which is often broken due to permanent updates - I used MKS repository as I know it is stable for MKS boards
So I am not sure they are at the same level of code - a good example is temperatures reporting ...

So Marlin Color UI is not related at all with MKS - So I do not know if MKS protocol can be enabled in that case  - better ask MKS or Marlin

**Reply by ColdBeer72** [reply]
You're wrong. I use MKS repository with LVGL UI

**Reply by luc-github** [reply]
Then something is broken in MKS repository - I used december version - if it can help - but as you can see your Marlin version is not giving temperatures - so issue is in Marlin part, MKS repository as now you say, but not ESP3D - I suggest you to raise issue to them

**Reply by ColdBeer72** [reply]
That's the point, Luc!!!

My intention is to concentrate most possible knowledge about ESP3D and MKS issues to could give them the truth information, not just issues. I want they understand that collaborating with ESP3D could give them a step ahead the competence.

That's why I just want to clarify those differences and every test. ;-)

**Reply by themacboy** [reply]
> 
> 
> Then something is broken in MKS repository - I used december version - if it can help - but as you can see your Marlin version is not giving temperatures - so issue is in Marlin part, MKS repository as now you say, but not ESP3D - I suggest you to raise issue to them

@ColBeer72
I use MKS repository but with the LVGL disabled. ANd I get good feedbask from board. hehehe

**Comment by LLHCrack** [comment]
I can't get any feedback on M114 to get location information on esp3d. M27 can get feedback on the printing progress, but it's all 0.  marlin2.0.9.3   robin nano v1.2

**Reply by luc-github** [reply]
yes Marlin MKS code filter M114 like several others as mentionned above,  about M27 I remember it is a bug in MKS Marlin code

**Comment by ETE-Design** [comment]
MKS Robin Lite - Wifi Module Direct on mainboard... No TFT just a regular LCD12864... Will it be possible to use Fast Transfer Protocol? Using Marlin Bugfix 2.0.x from Marlin's Github...

**Reply by luc-github** [reply]
does MKS wifi Code is present with this configuration ?
https://github.com/MarlinFirmware/Marlin/tree/2.0.x/Marlin/src/lcd/extui/mks_ui

**Reply by ETE-Design** [reply]
@luc-github So only a feature possible to use if you use Marlin from MKS Github?

**Reply by luc-github** [reply]
no the code is already in Marlin repository  for a while the link I gave is Marlin repository, did you checked?

**Reply by ETE-Design** [reply]
@luc-github I checked, there is some MKS Files and Wifi files... Remember that an old Version of MKS Own Firmware even had a menu in the LCD where you could search for network and connect it... But back to the case, Fast Transfer should work with LCD12864 if the MKS Wifi Modul is an addon module direct on the mainboard?

**Reply by luc-github** [reply]
so you know better Marlin than me, I do not use all config and I do not check code more than I need.
if the code I linked you is used then mks protocol is used if not I do not know what must be done on Marlin side to enable it
